jupyter_server/tests/test_firewall.py

```diff
@@ -0,0 +1,318 @@
+                        "name": "tags",
+                        "in": "query",
+                        "description": "Tags to filter by",
+                        "required": True,
+                        "style": "form",
+                        "schema": {"type": "array", "items": {"type": "string"}},
+                    }
+                ],
+                "responses": {
+                    "200": {
+                        "description": "successful operation",
+                    }
+                },
+            }
+        },
+        "/pet/{petId}": {
+            "get": {
+                "parameters": [
+                    {
+                        "name": "petId",
+                        "in": "path",
+                        "description": "ID of pet to return",
+                        "required": True,
+                        "schema": {"type": "integer", "format": "int64"},
+                    }
+                ],
+                "responses": {"200": {"description": ""}},
+            },
+        },
+    },
+    "components": {
+        "schemas": {
+            "Pet": {
+                "required": ["name"],
+                "type": "object",
+                "properties": {
+                    "id": {"type": "integer", "format": "int64"},
+                    "name": {"type": "string", "example": "doggie"},
+                },
+            },
+        }
+    },
+}
+
+spec2 = {
+    "openapi": "3.0.1",
+    "info": {"title": "Test specs", "version": "0.0.1"},
+    "paths": {
+        "/pet": {
+            "put": {
+                "responses": {"400": {"description": ""}},
+            }
+        },
+        "/user/{anyId}": {
+            "get": {
+                "parameters": [
+                    {
+                        "name": "anyId",
+                        "in": "path",
+                        "required": True,
+                        "schema": {"type": "string"},
+                    }
+                ],
+                "responses": {"200": {"description": ""}},
+            }
+        },
+    },
+}
+
+
+def format_id(val):
+    if isinstance(val, HTTPServerRequest):
+        return f"{val.method}-{val.uri}-{val.body.decode('utf-8') or None}"
+
+
+@pytest.mark.parametrize(
+    "base_url, server_request, expected",
+    (
+        (
+            "/",
+            HTTPServerRequest(
+                "put",
+                "/pet",
+                body=b'{"name":"puppy"}',
+                headers=HTTPHeaders({"Content-Type": "application/json"}),
+                host="localhost:8888",
+            ),
+            True,
+        ),
+        (
+            "/dummy/base_url/",
+            HTTPServerRequest(
+                "put",
+                "/dummy/base_url/pet",
+                body=b'{"name":"puppy"}',
+                headers=HTTPHeaders({"Content-Type": "application/json"}),
+                host="localhost:8888",
+            ),
+            True,
+        ),
+        # No body
```

Comment:
these comments are money for describing the _state of mind_ of the developer (in this case, you)... maybe put them into the test signature as `message` so they would actually show up when things fail